### DISCUSSION TITLE
Assigning permeability by block then pass it to the element-based compute permeability object [Materials][PorousFlow]

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#19340] [Q&A General] [answered]
Hi guys,

So I can assign permeability or any material property base on the elementID #18542 

But now, some of my projects are given permeability base on the block_ID, I'm wondering if it is possible, if yes, how to assign permeability based on the block_ID then pass it to my own element-based compute permeability object?

I'm thinking of something as follows

1. assign perm base on block ID using `PorousFlowPermeabilityConst`

```
  [permeability_A]
    type = PorousFlowPermeabilityConst
    block = A
    permeability = '1E-14 0 0   0 1E-14 0   0 0 1E-14'
  []

  [permeability_B]
    type = PorousFlowPermeabilityConst
    block = B
    permeability = '5E-14 0 0   0 5E-14 0   0 0 5E-14'
  []
```

2. store it to `AuxVariable/Auxkernel` and keep it constant throughout the simulation (not changing with time)
- Still not sure how to do this??
- maybe use `PorousFlowPropertyAux` then execute_on: initial? or only at t=1?

3. Then pass that `AuxVariable` to my own element-based compute permeability object

Please let me know what do you think?

Thank you

Kind regards,
Traiwit

**Comment by GiudGiud** [comment, marked as answer]
Hello

1) That same object I added can do block_id as well (and node id, and nearest-neighbor)
It just needs to be extended to write a tensor??
What you wrote seems to work though.

2) If it s a material property there's a MaterialPropertyAux to move it to an AuxVariable
There are declinations of this for vectors, tensors etc

3) Why an auxvariable specifically?
What does `PorousFlowPermeabilityConst` define? a matprop? 

Guillaume

**Reply by Traiwit** [reply]
> Hello
> 
> 1. That same object I added can do block_id as well (and node id, and nearest-neighbor)
>    It just needs to be extended to write a tensor??
>    What you wrote seems to work though.
> 2. If it s a material property there's a MaterialPropertyAux to move it to an AuxVariable
>    There are declinations of this for vectors, tensors etc
> 3. Why an auxvariable specifically?
>    What does `PorousFlowPermeabilityConst` define? a matprop?
> 
> Guillaume

Hi @GiudGiud 

1, yes that works fine

2. but it is defined at different blocks (with different object name [perm_1], [perm_2] ... [perm_n]) now sure how to put them together into 1 AuxVariable/Auxkernel, I found that https://mooseframework.inl.gov/source/auxkernels/PorousFlowPropertyAux.html allows you to do that.

3. PorousFlowPermeabilityConst is a matprop, but yeh if I don't park it somewhere (AuxVariable), it will be overwritten by my own compute perm (let's call it `PorousFlowPermeabilityConstCSV` for now)

**Reply by GiudGiud** [reply]
you can block restrict things. For example you can define material properties in one block at a time with the same name. 
Or you can use a MaterialAuxKernel to move a material property with a given name to an auxvariable in a single block, and do that N times to get all them in the same auxvariable

Umm that s annoying. Does the porous flow module not allow you to use different names? Then there s no concern about overwriting

**Reply by Traiwit** [reply]
Hi @GiudGiud, I almost got what I want to work, I will explain it when it's finished.

I now have a minor issue now, so `PorousFlowPropertyAux` doesn't store permeability in a tensor format (or vector format), it only gives me 1 permeability per element.

```
  [./permeability_bottom]
    type = PorousFlowPermeabilityConst
    block = TETS_BOTTOM_TET4
    permeability = '2e-9 0 0 0 3e-9 0 0 0 4e-9'
  [../]

  [./permeability_top]
    type = PorousFlowPermeabilityConst
    block = TETS_TOP_TET4
    permeability = '1e-9 0 0 0 2e-9 0 0 0 3e-9'
  [../]

  [./permeability_wedge]
    type = PorousFlowPermeabilityConst
    block = ALL_WDGS_ELSET_PRISM6
    permeability = '3e-9 0 0 0 2e-9 0 0 0 3e-9'
  [../]
```

 ```
 [./perm_map]
  variable = 'perm_map'
  type = PorousFlowPropertyAux
  property = permeability
  [../]
```

is there any other AuxKernel that I should look into? @WilkAndy is our PorousFlow expert, probs know about this?

Thank you!

**Comment by cpgr** [comment]
I'm not really sure why you need to define a permeability based on block ids and then a different permeability based on element ids? 

The problem you will have with the `PorousFlowPropertyAux` AuxKernel is that the AuxVariable has to be a monomial, so a single real value. You would need one for each non-zero component of the permeability tensor (probably three values needed). You can pick out the value using the `row` and `column` params (like `MaterialRealTensorValueAux`).

**Reply by Traiwit** [reply]
> I'm not really sure why you need to define a permeability based on block ids and then a different permeability based on element ids?
> 
> The problem you will have with the `PorousFlowPropertyAux` AuxKernel is that the AuxVariable has to be a monomial, so a single real value. You would need one for each non-zero component of the permeability tensor (probably three values needed). You can pick out the value using the `row` and `column` params (like `MaterialRealTensorValueAux`).

Ah I see, in this case, I should have 6 AuxVariables/AuxKernels then if I want to store all the perm tensors? 

a quick note about the element ids, I need that because the perm I will pass into the simulation will be calculated based on local damage, and the damage file which is obtained from Abaqus is element-based.
